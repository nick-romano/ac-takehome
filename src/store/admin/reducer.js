

export const reducer = (state, action) => {
    switch(action.type) {
        case 'setUser':
            return {
                ...state,
                user: action.user
            }
        default:
            throw new Error("unknown action dispatched to admin reducer");
    }
}
