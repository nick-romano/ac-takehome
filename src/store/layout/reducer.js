
export const reducer = (state, action) => {
    switch(action.type) {
        case "setAdminMenuOpen":
            return {
                ...state,
                adminMenuOpen: action.menuOpen
            }
        case "setSideMenuSelection": {
            return {
                ...state,
                sideMenuSelection: action.id
            }
        }
        case "setShelvesMenuExpanded":
            return {
                ...state,
                shelvesMenuExpanded: action.shelvesMenuExpanded
            }
        case "setAddShelfModalOpen":
            return {
                ...state,
                addShelfModalOpen: action.addShelfModalOpen
            }
        case "setNextPage":
            return {
                ...state,
                nextPage: action.nextPage
            }
        case "setChangeUserModalOpen":
            return {
                ...state,
                changeUserModalOpen: action.changeUserModalOpen
            }
        default:
            throw new Error("Unknown action type provided to admin reducer");
    }
}

