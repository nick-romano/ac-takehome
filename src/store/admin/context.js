import { createContext, useReducer } from "react";
import { reducer } from "./reducer";


export const initialState = {
    user: "blacklight"
};

const AdminContext = createContext(initialState);

const AdminProvider = ({ children }) => {

    const [state, dispatch] = useReducer(reducer, initialState);
    
    return (
        <AdminContext.Provider
            value={{
                ...state,
                dispatch
            }}
        >
            {children}
        </AdminContext.Provider>
    )
};

export { AdminProvider, AdminContext };





