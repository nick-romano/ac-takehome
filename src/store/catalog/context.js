import { createContext, useReducer } from "react";
import { reducer } from "./reducer";

export const initialState = {  
};

const CatalogContext = createContext(initialState); 

const CatalogProvider = ({ children }) => {

    const [state, dispatch] = useReducer(reducer, initialState);

    return (
        <CatalogContext.Provider
            value={{
                shelves: state,
                dispatch
            }}
        >
            {children}
        </CatalogContext.Provider>
    )

};

export { CatalogContext, CatalogProvider };