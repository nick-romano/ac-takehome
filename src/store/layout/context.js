import { createContext, useReducer } from "react";
import { reducer } from "./reducer";


export const initialState = {
    adminMenuOpen: false,
    shelvesMenuExpanded: true,
    sideMenuSelection: undefined,
    addShelfModalOpen: false,
    perPage: 25,
    nextPage: undefined,
    changeUserModalOpen: false
};

const LayoutContext = createContext(initialState);

const LayoutProvider = ({ children }) => {

    const [state, dispatch] = useReducer(reducer, initialState);
    
    return (
        <LayoutContext.Provider
            value={{
                ...state,
                dispatch
            }}
        >
            {children}
        </LayoutContext.Provider>
    )
};

export { LayoutProvider, LayoutContext };





