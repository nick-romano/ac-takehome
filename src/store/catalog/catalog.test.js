import { render } from '@testing-library/react';
import { CatalogProvider, initialState } from './context';
import { reducer, shelfIdCounter } from './reducer';

let state;

const getPrevious = () => `shelf-${shelfIdCounter - 1}`

describe('testing catalog store', () => {

    afterEach(() => { state = initialState });

    it('should render with children', () => {
        const { asFragment, getByText } = render(
        <CatalogProvider><div>TESTING!</div></CatalogProvider>
        )

        expect(getByText('TESTING!')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    })

    it('should update state to include a new shelf when create shelf is called', () => {
        
        state = reducer(state, {
            type: "createShelf",
            name: "TEST_SHELF"
        });

        const previous = `shelf-${shelfIdCounter - 1}`

        expect(state).toEqual(
            expect.objectContaining({
                [previous] : expect.objectContaining({
                    id: previous,
                    name: "TEST_SHELF",
                    records: []
                })
            })
        );
    });

    it('should update state to include delete shelf when delete shelf is called', () => {
        // add shelf;
        state = reducer(state, {
            type: "createShelf",
            name: "TEST_SHELF"
        });

        // make sure we have shelf-previous;
        expect(state).toEqual(
            expect.objectContaining({
                [getPrevious()]: expect.anything()
            })
        );

        // remove shelf
        state = reducer(state, {
            type: "deleteShelf",
            id: getPrevious()
        });

        // expect to not see previous shelf, because it was deleted
        expect(state).toEqual(
            expect.not.objectContaining({
                [getPrevious()]: expect.anything()
            })
        )
    });

    it('should rename a shelf', () => {
        // create shelf
        state = reducer(state, {
            type: "createShelf",
            name: "TEST_SHELF1"
        });
        // rename latest shelf
        state = reducer(state, {
            type: "renameShelf",
            id: getPrevious(),
            name: "Cool Shelf"
        });

        expect(state[getPrevious()].name).toBe("Cool Shelf")
    });

    it('should add a record to the shelf', () => {
        // create shelf
        state = reducer(state, {
            type: "createShelf",
            name: "TEST_SHELF1"
        });

        // create shelf
        state = reducer(state, {
            type: "addRecordToShelf",
            shelfId: getPrevious(),
            recordId: "76035"
        });

        expect(state[getPrevious()].records).toEqual(["76035"]);
    });

    it('should remove a record from the shelf', () => {
        // create shelf
        state = reducer(state, {
            type: "createShelf",
            name: "TEST_SHELF1"
        });

        // create shelf
        state = reducer(state, {
            type: "addRecordToShelf",
            shelfId: getPrevious(),
            recordId: "76035"
        });

        state = reducer(state, {
            type: "removeRecordFromShelf",
            shelfId: getPrevious(),
            recordId: "76035",
        })

        expect(state[getPrevious()].records).toEqual([]);
    });

    it('should reset the state', () => {
         // create shelf
         state = reducer(state, {
            type: "createShelf",
            name: "TEST_SHELF1"
        });

        // create shelf
        state = reducer(state, {
            type: "addRecordToShelf",
            shelfId: getPrevious(),
            recordId: "76035"
        });

        state = reducer(state,{
            type: "resetState"
        })

        expect(state).toEqual(initialState)
    })

    it('should throw an error with unknown action type', () => {
        expect(() => reducer(initialState, {
            type: "randomType",
            user: "TEST"
        })).toThrow()
    })
});