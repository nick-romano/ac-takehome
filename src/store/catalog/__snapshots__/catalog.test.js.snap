// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing catalog store should render with children 1`] = `
<DocumentFragment>
  <div>
    TESTING!
  </div>
</DocumentFragment>
`;
